### DISCUSSION TITLE
Visualize a remotely hosted state machine

### DISCUSSION THREAD

**Discussion by zacharyweidenbach** [statelyai/xstate#3426] [Q&A]
The state machine visualizer is a great way to see the inner workings of a state machine. In the tutorials I've seen, the visualizer functions as a local dev tool, but is it possible to see the visualizer of a state machine hosted in some remote server somewhere?

One approach would be to hook into the `onChange` or `onTransition` of the machine and have it send its full state via websocket to a connected client. The client could take the received state, and feed it into the visualizer, and the output of this visualizer could be rendered in a UI. The missing piece is how to access the visualizer library directly. The `@xstate/inspect` package seems to only output to a URL at `https://stately.ai/viz`. Is there a way to directly handle the outputted HTML/CSS from the visualizer? Or is the rendering portion of the visualizer closed source?

**Comment by zacharyweidenbach** [comment]
After some more digging, it looks like there is experimental support for this right now https://github.com/statelyai/xstate/discussions/1795#discussioncomment-252575

And after looking at the xstate-viz github package, it looks like they provide a means for self-hosting the visualizer if you don't want to delegate this to the stately.ai host https://github.com/statelyai/xstate-viz#installation

**Reply by Andarist** [reply]
The new viz also has support for the `server` query param: https://github.com/statelyai/xstate-viz/releases/tag/v0.7.0

And I guess the rest of your questions have been answered since you have found the xstate-viz repo: https://github.com/statelyai/xstate-viz
